
#########################
Mutants Browser DashBoard
#########################
of the `Plant Breeding and Genetics Laboratory, FAO/IAEA Joint Division <http://www-naweb.iaea.org/nafa/pbg/index.html>`_


.. image:: https://img.shields.io/conda/dn/bioconda/snakemake.svg?label=Bioconda
    :target: https://bioconda.github.io/recipes/snakemake/README.html

.. image:: https://img.shields.io/pypi/pyversions/snakemake.svg
    :target: https://www.python.org

.. .. raw:: html
          <span class="__dimensions_badge_embed__" data-doi="https://doi.org/10.1093/bioinformatics/bts480" data-legend="always" data-style="large_rectangle"></span><script async src="https://badge.dimensions.ai/badge.js" charset="utf-8"></script>



*******
Purpose
*******

This DashBoard provides a search interface for DNA variation. The intended purpose is a searchable catalogue of mutations and mutant individuals that is queried by genes-of-interest or chromosome/position ranges. 
The user will need to provide a bcf/vcf file generated by SnpEff (http://snpeff.sourceforge.net/). 
The DashBoard will use the annotation information from this file. We provide a parser that will generate the data tables that the DashBoard will need. The user can add passport information to the individuals and custom chromosome identifiers. 

-----------------
For the impatient
----------------- 

.. code-block:: bash

    $ conda env create -f envs/condaenv.yaml
    $ conda activate Dashboard
    $ bash utils/vcf_to_datatables.sh ../input.snpeff.vcf.gz 
    $ python mutants_dashboard.py 



************************
Setting up the DashBoard
************************

--------------------------------
Installing the conda environment
--------------------------------


The DashBoard will need to run in a conda environment. 

Create the environment defined in **envs/condaenv.yaml** like so:

.. code-block:: bash

    $ conda env create -f envs/condaenv.yaml

Once created, activate the environment **Dashboard** like so:

.. code-block:: bash  

    $ conda activate Dashboard


-------------------------
Creating the data tables
-------------------------

Prerequisite is a vcf/bcf file that has been run through SnpEeff to annotated the variant's effects (http://snpeff.sourceforge.net/). 
On such annotated vcf/bcf file run the parser script like so:

.. code-block:: bash

    $ bash utils/vcf_to_datatables.sh <path to input snpeff vcf/bcf>


This will generate all required data tables.

* snpsiftdata.tab
* genotype_data.tab
* passport.tab
* chromosome_name_mapping.tab

The user now has the option to edit **passport.tab** in order to add details to individual samples. In **chromosome_name_mapping.tab** custom chromosome identifiers can be provided in the "Chromosome" column. 
Note that subsequent runs of the parser (**vcf_to_datatables.sh**) will overwrite the existing files. Details on how the tables are generated can be found further below.

----------------------
Creating the Dashboard
----------------------

With the data tables generated and in **./data**, run the following command to create the DashBoard:

.. code-block:: bash

    $ python mutants_dashboard.py

Upon startup the dashboard will report where its running:

.. code-block:: bash

    Running on http://127.0.0.1:8050/
    Debugger PIN: 383-685-305
    * Serving Flask app "mutants_dashboard" (lazy loading)
    * Environment: production
    WARNING: This is a development server. Do not use it in a production deployment.


In the above/standard case, point your browser to URL **http://127.0.0.1:8050/** to view the DashBoard.
The DashBoard's display name can be configured by editing the "institution" and "tool" variables in **mutants_dashboard.py**.   


.. _Details on generating the data tables:

*************************************
Details on generating the data tables
*************************************

We kept the extraction of relevant information from the vcf/bcf file into the data tables quite simple and use only standard software tools.
All commands are executed when running **./utils/vcf_to_datatables.sh**. We provide details below for transparency. 
The commands are easily tested and adjusted to meet specific needs. Example data tables are provided in **./data_example** for reference. 


-------------
SnpSift Data (snpsiftdata.tab)
-------------

This file is created from the <inputfile.vcf/bcf> by utils/vcf_to_datatables.sh like so:

.. code-block:: python

    bcftools view <inputfile.vcf/bcf> | grep -v "start_retained_variant" | \
    $CONDA_PREFIX/share/snpsift-*/scripts/vcfEffOnePerLine.pl | \
    SnpSift extractFields -e "NA" - "ANN[*].GENE" "ANN[*].DISTANCE" CHROM POS ID REF ALT TYPE "ANN[*].IMPACT" "ANN[*].EFFECT" "ANN[*].FEATURE" "ANN[*].FEATUREID" "ANN[*].BIOTYPE" "ANN[*].RANK" \
    > data/snpsiftdata.tab



It uses snpEff/SnpSifts own functionality and scripts to extract the relevant information per variant and effect.
We are excluding lines with effect type "start_retained_variant", because SnpSift does not seem to understand this effect type.

-----------------------------------------------------
Chromosome Name Mapping (chromosome_name_mapping.tab)
-----------------------------------------------------

This file is created from the <inputfile.vcf/bcf> by utils/vcf_to_datatables.sh like so:

.. code-block:: bash

    printf "Contig\\tChromosome\n" > data/chromosome_name_mapping.tab
    bcftools view -h <inputfile.vcf/bcf> | grep "##cont"| \
    awk -F "=|," '{print $3 "\t" $3}' >> data/chromosome_name_mapping.tab

Information on chromosome names is extracted from the vcf/bcf file and recorded twice (in 2 columns), as "Contig" and "Chromosome". 
The "Contig" column must remain unchanged. By editing the "Chromosome" column the user has the option of mapping the "Contig" names to custom chromosome identifiers.  


---------------------------------
Genotype Data (genotype_data.tab)
---------------------------------

This file is created from the <inputfile.vcf/bcf> by utils/vcf_to_datatables.sh like so:

.. code-block:: bash

    CHROM_POS=$(printf "CHROM\\tPOS\\t");
    SAMPLE_NAMES=$(bcftools query -l <inputfile.vcf/bcf> | paste -s -d "\t" -)
    echo "$CHROM_POS$SAMPLE_NAMES"> data/genotype_data.tab
    bcftools view <inputfile.vcf/bcf> | bcftools query -f "%CHROM\t%POS[\t%GT]\n" >> data/genotype_data.tab

It simply extracts

----------------------------
Passport Data (passport.tab)
----------------------------

This table is initially populated with the sample names found in the vcf/bcf file in the "Sample-ID" column and "NA" in each of the data columns. The user has the option to edit this file to provide the relevant passport information for each of the samples. "Sample-ID" must not be edited. It must be a complete list of samples from the vcf/bcf file and must match the sample names. 
 

.. code-block:: bash

    printf "Sample-ID\\tPlant-ID\\tBranch-ID\\tVariety\\tGeneration\\tTreatment\\tDose\n" > data/passport.tab
    a=$(bcftools query -l <inputfile.vcf/bcf>)
    b="\tNA\tNA\tNA\tNA\tNA\tNA"
    for i in ${a[*]}; do
       echo -e $i$b >> data/passport.tab;
    done


Example **passport.tab** file (after manual editing):

      +-----------+-------------+-----------+-----------+------------+------------+--------+
      | Sample-ID | Plant-ID    | Branch-ID | Variety   | Generation | Treatment  | Dose   | 
      +===========+=============+===========+===========+============+============+========+
      | 1-C7      | Ca-2018-021 | NA        | Venetia   | M0         | Control    | NA     |
      +-----------+-------------+-----------+-----------+------------+------------+--------+
      | 1-D4      | Ca-2018-025 | NA        | Venetia   | M1         | EMS        | 2%     |
      +-----------+-------------+-----------+-----------+------------+------------+--------+
      | 1-E2      | Ca-2018-030 | NA        | Venetia   | M1         | Gamma      | 50 Gy  |
      +-----------+-------------+-----------+-----------+------------+------------+--------+

*********************
Copyright information
*********************

This Dashboard was developed by Anza Ghaffar and Norman Warthmann, 
Â© 2020 `Plant Breeding and Genetics Laboratory of the FAO/IAEA Joint Division <http://www-naweb.iaea.org/nafa/pbg/index.html>`_.
If you find this DashBoard useful and want to use in in your own research, please get in touch by emailing
n.warthmann@iaea.org. We are happy to provide an annotated vcf/bcf for testing.



